sympy/matrices/determinant.py

```diff
@@ -0,0 +1,977 @@
+        speed up calculation.
+
+    Examples
+    ========
+
+    >>> from sympy import Matrix
+    >>> from sympy.matrices import minor
+    >>> M = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+    >>> M.minor(1, 1)
+    -12
+    >>> minor(M, 1, 1)
+    -12
+
+    See Also
+    ========
+
+    minor_submatrix
+    cofactor
+    det
+    """
+
+    if not M.is_square:
+        raise NonSquareMatrixError()
+
+    return M.minor_submatrix(i, j).det(method=method, dotprodsimp=dotprodsimp)
+
+
+def _minor_submatrix(M, i, j):
+    """Return the submatrix obtained by removing the `i`th row
+    and `j`th column from ``M`` (works with Pythonic negative indices).
+
+    Parameters
+    ==========
+
+    i, j : int
+        The row and column to exclude to obtain the submatrix.
+
+    Examples
+    ========
+
+    >>> from sympy import Matrix
+    >>> from sympy.matrices import minor_submatrix
+    >>> M = Matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+    >>> M.minor_submatrix(1, 1)
+    Matrix([
+    [1, 3],
+    [7, 9]])
+    >>> minor_submatrix(M, 1, 1)
+    Matrix([
+    [1, 3],
+    [7, 9]])
+
+    See Also
+    ========
+
+    minor
+    cofactor
+    """
+
+    if i < 0:
+        i += M.rows
+    if j < 0:
+        j += M.cols
+
+    if not 0 <= i < M.rows or not 0 <= j < M.cols:
+        raise ValueError("`i` and `j` must satisfy 0 <= i < ``M.rows`` "
+                            "(%d)" % M.rows + "and 0 <= j < ``M.cols`` (%d)." % M.cols)
+
+    rows = [a for a in range(M.rows) if a != i]
+    cols = [a for a in range(M.cols) if a != j]
+
+    return M.extract(rows, cols)
+
+
+# The following are top level stand-alone interface functions which sympify the
+# matrix where needed (so it becomes immutable and returns immutable), otherwise
+# the implementations above do not sympify due to problems in other parts of the
+# codebase using matrices of unsympifiable objects. Technically could just
+# assign these via 'charpoly = _charpoly' where sympification is not needed but
+# this is clearer.
+
+def adjugate(M, method="berkowitz", dotprodsimp=None):
+    if isinstance(M, Expr):
+        raise NotImplementedError('adjugate for matrix expressions not supported yet')
+
+    return sympify(_adjugate(M, method=method, dotprodsimp=dotprodsimp))
+
+def charpoly(M, x='lambda', simplify=_simplify, dotprodsimp=None):
+    if isinstance(M, Expr):
+        raise NotImplementedError('charpoly for matrix expressions not supported yet')
+
+    return _charpoly(M, x=x, simplify=simplify, dotprodsimp=dotprodsimp)
+
+def cofactor(M, i, j, method="berkowitz", dotprodsimp=None):
+    return _cofactor(M, i, j, method=method, dotprodsimp=dotprodsimp)
+
+def cofactor_matrix(M, method="berkowitz", dotprodsimp=None):
+    return sympify(_cofactor_matrix(M, method=method, dotprodsimp=dotprodsimp))
+
+def det(M, method="bareiss", iszerofunc=None, dotprodsimp=None):
+    from .expressions import Determinant
+
+    if isinstance(M, Expr):
+        return Determinant(M).doit()

```

Comment:
`ImmutableMatrix` is a subclass of `Expr` but `Matrix` is not.

Reply:
That is weird, why is it a `MatrixExpr` when `ImmutableSparseMatrix` is `Basic`, is this a recent change? Doesn't matter, works anyway since `Determinant.doit` -> `_eval_determinant` -> `_det`, so no infinite recursion.

Reply:
Though it will lose the `method` and `dotprodsimp`...

Reply:
Actually, regardless of where the top level `det` function lives this check would have to be made the same way. There are not many top-level standalone functions in expressions currently, one of them was `det`, another is `diagonalize_vector` in `diagonal.py` which seems redundant with a top level `diagonalize` - this could be coalesced into one function. `inverse.py` does not have a top-level `inv` or `inverse` where that may be desirable? Or would that be too general?

Note that I am only mentioning these as examples related to the `det` question, not suggesting changing any of these here. As for `det`, I think it makes more sense for the top level `det` to be above `expressions` and delegate to that from where I put it. If you disagree just let me know and I will send it back there and change the docs.

Reply:
If this function receives those arguments then it should use them where appropriate. Otherwise it should not receive those arguments at all.

Another possibility is to pass them through `doit`.

Reply:
Note the check I currently have in `det` is:
```
    if isinstance(M, Expr) and \
            not isinstance (M, (ImmutableDenseMatrix, ImmutableSparseMatrix)):
        return Determinant(M).doit()
```
I could pass this on to `doit`, but then `doit` calls `_eval_determinant` and for something like `MatrixSymbol` this would not make sense until an actual matrix was substituted for it. In which case I would either have to store these parms in the `args` for later retrieval - creating different `Determinant` objects depending on what args these take even if they have the same first matrix symbol or object, or discard them and lose the information - which is what I believe the intent was of the top level `Determinant` object to not have to deal with these specifics.